subprojects/core/src/main/java/org/gradle/api/services/internal/DefaultBuildServicesRegistry.java

```diff
@@ -118,40 +122,82 @@ public SharedResource forService(BuildServiceProvider<?, ?> service) {
             int maxUsages = registration.getMaxParallelUsages().getOrElse(-1);
 
             if (maxUsages > 0) {
-                leaseRegistry.registerSharedResource(serviceName, maxUsages);
+                leaseRegistry.registerSharedResource(registration.getName(), maxUsages);
             }
-            return new ServiceBackedSharedResource(serviceName, maxUsages, leaseRegistry);
+            return new ServiceBackedSharedResource(registration.getName(), maxUsages, leaseRegistry);
         });
     }
 
+    @Override
+    public DefaultServiceRegistration<?, ?> findRegistration(Class<?> type, String name) {
+        return uncheckedCast(!StringUtils.isEmpty(name) ?
+            findByName(name) :
+            findByType(type)
+        );
+    }
+
     @Override
     @Nullable
     public BuildServiceRegistration<?, ?> findByName(String name) {
         return withRegistrations(registrations -> registrations.findByName(name));
     }
 
+    @Nullable
     @Override
-    public <T extends BuildService<P>, P extends BuildServiceParameters> Provider<T> registerIfAbsent(String name, Class<T> implementationType, Action<? super BuildServiceSpec<P>> configureAction) {
-        return withRegistrations(registrations -> {
-            BuildServiceRegistration<?, ?> existing = registrations.findByName(name);
-            if (existing != null) {
-                // TODO - assert same type
-                // TODO - assert same parameters
-                return uncheckedNonnullCast(existing.getService());
-            }
+    public BuildServiceRegistration<?, ?> findByType(Class<?> type) {
+        Collection<BuildServiceRegistration<?, ?>> results = withRegistrations(registrations ->
+            new ArrayList<>(
+                registrations.matching(it -> type.isAssignableFrom(((ProviderInternal) it.getService()).getType()))
+            )
+        );
+        if (results.size() > 1) {
```

Comment:
As discussed in the daily call, we ensure we fail early if we cannot unambiguously resolve by type